### ISSUE TITLE
Switch case after `or` pattern does not match

### ISSUE THREAD

**Issue by WiebeCnossen** [dotnet/roslyn#55668]
**Version Used**: 5.0.302

**Steps to Reproduce**:

Given this method:

```
    private static string GetTextBuggy(Position? position, Wrap wrap) {
      return position switch {
        not Position.First when wrap.Sub is Zero => "Not First and Zero",

        null => "Null",

        _ when wrap is {
          Sub: One or Two
        } => "One or Two",

        Position.Last => "Last",

        Position.First => "First",

        _ => "Other"
      };
    }
```

This is called with argument values `Position.First` and `Wrap { Sub = new Zero() }`.

**Expected Behavior**: The return value is "First"

**Actual Behavior**: The return value is "Other".

**Remarks**: The code works as expected when the `or` pattern is split into to cases:

```
    private static string GetTextWorkaround(Position? position, Wrap wrap) {
      return position switch {
        not Position.First when wrap.Sub is Zero => "Not First and Zero",

        null => "Null",

        _ when wrap is {
          Sub: One
        } => "Just one",

        _ when wrap is {
          Sub: Two
        } => "Just two",

        Position.Last => "Last",

        Position.First => "First",

        _ => "Other"
      };
    }
```

Complete project: [Switcher.zip](https://github.com/dotnet/roslyn/files/7000454/Switcher.zip)

**Comment by Halbjobbit** [comment]
The expected output is also returned when the first pattern is removed. When you remove the null-check it again returns different values which also are incorrect.

**Comment by jcouv** [comment]
Confirmed there's an issue. The following full program prints "Other" but should print "Fist":
```
class C
{
    public static void Main()
    {
        System.Console.Write(GetTextBuggy(Position.First, new Wrap { Sub = new Zero() }));
    }

    private static string GetTextBuggy(Position? position, Wrap wrap)
    {
        return position switch
        {
            not Position.First when wrap.Sub is Zero => "Not First and Zero",

            null => "Null",

            _ when wrap is
            {
                Sub: One or Two
            } => "One or Two",

            Position.Last => "Last",

            Position.First => "First",

            _ => "Other"
        };
    }
}

enum Position
{
    First,
    Last,
}

class Zero { }
class One { }
class Two { }

class Wrap
{
    public object? Sub;
}
```

Note that the problem doesn't repro after removing the wrapping (removing the `Wrap` type, modifying `GetTextBuggy` to take an `object?` and updating the `switch` accordingly).

**Comment by jcouv** [comment]
Here's what I found so far from looking at simplified example above.
In short, there are scenarios where `when` clauses will be duplicated into the decision DAG, but this can't work when the `when` clause includes label statements, because those label statements will get duplicated. 
Using a debug compiler, this triggers a compiler assertion (we've emitted the same label twice). 
Using a release compiler, some of the jumps will go to the wrong location, while results in the behavior described in OP (incorrect execution).

For that example, the DAG looks like this:
```
[0]: t0 == 0 ? [1] : [3]
[1]: when (wrap is { Sub: One or Two }) ? [6] : [2]
[2]: leaf <arm> `Position.First => "First"`
[3]: when (wrap.Sub is Zero) ? [7] : [4]
[4]: when (wrap is { Sub: One or Two }) ? [6] : [5]
[5]: leaf <arm> `_ => "Other"`
[6]: leaf <arm> `_ when wrap is { Sub: One or Two } => "One or Two"`
[7]: leaf <arm> `not Position.First when wrap.Sub is Zero => "Not First and Zero"`
```

Notice that:
1. states `[1]` and `[4]` duplicate the `when` condition `wrap is { Sub: One or Two }`,
2. that condition is complex (it involves labels, because of the `or` pattern),
3. states `[1]` and `[4]` use that condition to jump to different cases (`[6]`/`[2]` versus `[6]`/`[5]`, so it's important not to accidentally jump to a duplicated label in the wrong condition).
